File: udon_decompiler/analysis/transform/passes/collect_variables.py
from __future__ import annotations

from typing import Dict

from udon_decompiler.analysis.ir.nodes import (
    IRAssignmentStatement,
    IRBlock,
    IRBlockContainer,
    IRConstructorCallExpression,
    IRExpression,
    IRExpressionStatement,
    IRExternalCallExpression,
    IRFunction,
    IRHighLevelSwitch,
    IRIf,
    IRInternalCallExpression,
    IRLiteralExpression,
    IROperatorCallExpression,
    IRPropertyAccessExpression,
    IRStatement,
    IRSwitch,
    IRVariableDeclarationStatement,
    IRVariableExpression,
)
from udon_decompiler.analysis.transform.pass_base import (
    IILTransform,
    ILTransformContext,
)
from udon_decompiler.analysis.variable_identifier import Variable


class CollectVariables(IILTransform):
    """Collect referenced variables into ``function.variable_declarations``."""

    def run(self, function: IRFunction, context: ILTransformContext) -> None:
        _ = context

        seen_blocks: set[int] = set()
        collected: Dict[int, Variable] = {}

        self._collect_from_container(function.body, seen_blocks, collected)

        declarations = [
            IRVariableDeclarationStatement(variable=var, init_value=None)
            for _, var in sorted(collected.items(), key=lambda item: item[0])
        ]
        function.variable_declarations = declarations

    def _collect_from_container(
        self,
        container: IRBlockContainer,
        seen_blocks: set[int],
        collected: Dict[int, Variable],
    ) -> None:
        for block in container.blocks:
            self._collect_from_block(block, seen_blocks, collected)

    def _collect_from_block(
        self,
        block: IRBlock,
        seen_blocks: set[int],
        collected: Dict[int, Variable],
    ) -> None:
        block_id = id(block)
        if block_id in seen_blocks:
            return
        seen_blocks.add(block_id)

        for statement in block.statements:
            self._collect_from_statement(statement, seen_blocks, collected)

    def _collect_from_statement(
        self,
        statement: IRStatement,
        seen_blocks: set[int],
        collected: Dict[int, Variable],
    ) -> None:
        if isinstance(statement, IRAssignmentStatement):
            self._collect_variable(statement.target, collected)
            self._collect_from_expression(statement.value, collected)
            return

        if isinstance(statement, IRExpressionStatement):
            self._collect_from_expression(statement.expression, collected)
            return

        if isinstance(statement, IRIf):
            self._collect_from_expression(statement.condition, collected)
            self._collect_from_statement(
                statement.true_statement,
                seen_blocks,
                collected,
            )
            if statement.false_statement is not None:
                self._collect_from_statement(
                    statement.false_statement,
                    seen_blocks,
                    collected,
                )
            return

        if isinstance(statement, IRBlock):
            self._collect_from_block(statement, seen_blocks, collected)
            return

        if isinstance(statement, IRBlockContainer):
            self._collect_from_container(statement, seen_blocks, collected)
            return

        if isinstance(statement, IRSwitch):
            self._collect_from_expression(statement.index_expression, collected)
            return

        if isinstance(statement, IRHighLevelSwitch):
            self._collect_from_expression(statement.index_expression, collected)
            for section in statement.sections:
                self._collect_from_block(section.body, seen_blocks, collected)
            return

    def _collect_from_expression(
        self,
        expression: IRExpression,
        collected: Dict[int, Variable],
    ) -> None:
        if isinstance(expression, IRVariableExpression):
            self._collect_variable(expression.variable, collected)
            return

        if isinstance(
            expression,
            (
                IRExternalCallExpression,
                IRPropertyAccessExpression,
                IRConstructorCallExpression,
                IROperatorCallExpression,
            ),
        ):
            for argument in expression.arguments:
                self._collect_from_expression(argument, collected)
            return

        if isinstance(expression, (IRLiteralExpression, IRInternalCallExpression)):
            return

    def _collect_variable(
        self,
        variable: Variable,
        collected: Dict[int, Variable],
    ) -> None:
        # Exclude pseudo receiver names from declaration emission.
        if variable.name == "this" or "." in variable.name:
            return

        if variable.address not in collected:
            collected[variable.address] = variable

File: src/decompiler/transform/passes/collect_variables.rs
use std::collections::{BTreeSet, HashMap};

use crate::decompiler::Result;
use crate::decompiler::ir::{
    IrAssignmentStatement, IrExpression, IrExpressionStatement, IrFunction, IrIf, IrStatement,
    IrSwitch, IrVariableDeclarationStatement,
};
use crate::decompiler::transform::pass_base::{ITransform, TransformContext};
use crate::str_constants::{
    SYMBOL_PREFIX_THIS, SYMBOL_THIS, SYMBOL_THIS_GAME_OBJECT, SYMBOL_THIS_TRANSFORM,
};

pub struct CollectVariables;

impl ITransform for CollectVariables {
    fn run(&self, function: &mut IrFunction, context: &mut TransformContext<'_, '_>) -> Result<()> {
        let mut used = BTreeSet::<u32>::new();
        for block in &function.body.blocks {
            for statement in &block.statements {
                collect_statement_variables(statement, &mut used);
            }
        }

        let all_vars = context
            .program_context
            .decompile_context
            .variables
            .variables
            .iter()
            .map(|x| (x.address, x))
            .collect::<HashMap<_, _>>();

        let mut declarations = Vec::<IrVariableDeclarationStatement>::new();
        for address in used {
            let Some(variable) = all_vars.get(&address) else {
                continue;
            };

            if variable.exported || is_this_like(variable.name.as_str()) {
                continue;
            }

            let init_value = Some(crate::decompiler::ir::IrLiteralExpression {
                value: variable.init_value.clone(),
                type_hint: variable.type_name.clone(),
            });

            declarations.push(IrVariableDeclarationStatement {
                variable_address: variable.address,
                init_value,
            });
        }

        function.variable_declarations = declarations;
        Ok(())
    }
}

fn collect_statement_variables(statement: &IrStatement, used: &mut BTreeSet<u32>) {
    match statement {
        IrStatement::Assignment(IrAssignmentStatement { target, value }) => {
            collect_assignment_target_variables(target, used);
            collect_expression_variables(value, used);
        }
        IrStatement::Expression(IrExpressionStatement { expression }) => {
            collect_expression_variables(expression, used);
        }
        IrStatement::If(IrIf {
            condition,
            true_statement,
            false_statement,
        }) => {
            collect_expression_variables(condition, used);
            collect_statement_variables(true_statement, used);
            if let Some(false_statement) = false_statement.as_ref() {
                collect_statement_variables(false_statement, used);
            }
        }
        IrStatement::Switch(IrSwitch {
            index_expression, ..
        }) => {
            collect_expression_variables(index_expression, used);
        }
        IrStatement::Block(block) => {
            for statement in &block.statements {
                collect_statement_variables(statement, used);
            }
        }
        IrStatement::BlockContainer(container) => {
            for block in &container.blocks {
                for statement in &block.statements {
                    collect_statement_variables(statement, used);
                }
            }
        }
        IrStatement::HighLevelSwitch(switch_stmt) => {
            collect_expression_variables(&switch_stmt.index_expression, used);
            for section in &switch_stmt.sections {
                for statement in &section.body.statements {
                    collect_statement_variables(statement, used);
                }
            }
        }
        IrStatement::HighLevelWhile(while_stmt) => {
            if let Some(condition) = while_stmt.condition.as_ref() {
                collect_expression_variables(condition, used);
            }
            for block in &while_stmt.body.blocks {
                for statement in &block.statements {
                    collect_statement_variables(statement, used);
                }
            }
        }
        IrStatement::HighLevelDoWhile(do_while_stmt) => {
            collect_expression_variables(&do_while_stmt.condition, used);
            for block in &do_while_stmt.body.blocks {
                for statement in &block.statements {
                    collect_statement_variables(statement, used);
                }
            }
        }
        _ => {}
    }
}

fn collect_assignment_target_variables(target: &IrExpression, used: &mut BTreeSet<u32>) {
    match target {
        IrExpression::Variable(variable_expression) => {
            used.insert(variable_expression.address);
        }
        IrExpression::PropertyAccess(property_access) => {
            for arg in &property_access.arguments {
                collect_expression_variables(arg, used);
            }
        }
        IrExpression::ArrayAccess(array_access) => {
            for arg in &array_access.arguments {
                collect_expression_variables(arg, used);
            }
        }
        _ => {}
    }
}

fn collect_expression_variables(expression: &IrExpression, used: &mut BTreeSet<u32>) {
    match expression {
        IrExpression::Variable(variable_expression) => {
            used.insert(variable_expression.address);
        }
        IrExpression::ExternalCall(call) => {
            for arg in &call.arguments {
                collect_expression_variables(arg, used);
            }
        }
        IrExpression::PropertyAccess(call) => {
            for arg in &call.arguments {
                collect_expression_variables(arg, used);
            }
        }
        IrExpression::ArrayAccess(call) => {
            for arg in &call.arguments {
                collect_expression_variables(arg, used);
            }
        }
        IrExpression::ConstructorCall(call) => {
            for arg in &call.arguments {
                collect_expression_variables(arg, used);
            }
        }
        IrExpression::OperatorCall(call) => {
            for arg in &call.arguments {
                collect_expression_variables(arg, used);
            }
        }
        _ => {}
    }
}

fn is_this_like(name: &str) -> bool {
    name == SYMBOL_THIS
        || name == SYMBOL_THIS_TRANSFORM
        || name == SYMBOL_THIS_GAME_OBJECT
        || name.starts_with(SYMBOL_PREFIX_THIS)
}
